### ISSUE TITLE
Find all missing keys passed through translate pipe in the whole project (unit test)

### ISSUE THREAD

**Issue by AlexeyGalimov** [ngx-translate/core#1152]
I want to write the ONE unit test for the whole project (not for one specified component) in order to check if there are any missing keys (in my .json file with all translations), which I pass through translation pipe ANYWHERE in the project. For example:

`<span>{{ 'FOO' | translate }}</span>`

If FOO key doesn`t exist in .json file, then the test should fail and (optionally) somehow return the list of missing keys. Is it possible to implement this feature in this way? Or at least in other comparable ways?

**Comment by Wisdomb33r** [comment]
@AlexeyGalimov We did something similar in one of our project (except we print the missing translations in a specific debug route rather than while unit testing).

First of all, you should have a look at `MissingTranslationHandler` class of the framework, a handler called whenever a key is missing :

```TypeScript
export class MyMissingTranslationHandler implements MissingTranslationHandler {
  handle(params: MissingTranslationHandlerParams): any {
    const missingKey = params.key;
    this.missingTranslations[missingKey] = true; // or any structure holding missing translations
    return `?${missingKey}?`;
  }
}
```

And provide it within your root module with `{provide: MissingTranslationHandler, useClass: MyMissingTranslationHandler},`

In your unit tests, you can use an `afterAll` call which gets the `MissingTranslationHandler` from TestBed (it's accessible from the injector) and retrieve the missing translation keys. How you save this is up to you (console warn is an easy start, implement a more complex behavior such as file write if that suits your need).

For your point about writing a single unit test, I don't think there is any magic way of doing this. `ngIf` directives in your code are for sure `hiding` some of the less-used translations (like awful error messages or the likes), and some of these messages might appear only under very specific circumstances with very specific data. If you want to be sure to pass through ALL the translations (pipe or not) in your application, I suppose the only way is to have a 100% unit test cover. Which seems unrealistic unless you write code sending rockets to the moon or flying planes.

**Comment by AlexeyGalimov** [comment]
@Wisdomb33r Yes, of course I know about `MissingTranslationHandler`, thanks. But if I understood correctly before, it can be used when I work in the application and come across with missing translations which are then passed to the `MissingTranslationHandler`. And during unit tests, I can't just use the `MissingTranslationHandler`, because I also need to use the `TranslateModule` (and then the unit test will become non-unit) and provide it my own implementation of the `MissingTranslationHandler`. Am I right?

Thus, you suggest me using the following:
1) Implement and provide the `MissingTranslationHandler`
2) Work for a while in the application and collect missing translation keys
3) Run the unit test, where we get the `MissingTranslationHandler` from `TestBed` and retrieve the missing translation keys

As for point 1 - How to provide `MissingTranslationHandler` in `TestBed` without import all `TranslateModule` or to import/mock it correctly.
As for point 2 - if I and my teammates will want to work on it on different machines, we will need to work in app again, and it is unlikely that each of us will get the same result. If we will save the missing translation keys into a file and add it to the git, the list of keys may be irrelevant, especially if the test did not run long time.

Could you give a little more explanation on how best to do my issue? Maybe I just don’t know some approaches and mechanisms of `MissingTranslationHandler` or of other? Thanks in advance!

**Comment by svoboda-rabstvo** [comment]
@AlexeyGalimov you can check missing translation keys with [`ngx-translate-lint`](https://github.com/svoboda-rabstvo/ngx-translate-lint)
